File: 프로그래머스/2/148652. 유사 칸토어 비트열/유사 칸토어 비트열.java
class Solution {
    static String s = "11011";

    public int solution(int n, long l, long r) {
        int answer = dfs(n, r) - dfs(n, l - 1);
        return answer;
    }

    private static int dfs(int n, long k) {
        int count = 0;

        if (n == 1) {
            for (int i = 0; i < k; i++) {
                if (s.charAt(i) == '1') {
                    count++;
                }
            }
            return count;
        }

        long a = k / (long) Math.pow(5, n - 1);
        long b = k % (long) Math.pow(5, n - 1);

        if (a <= 1) {
            count = (int) (Math.pow(4, n - 1) * a + dfs(n - 1, b));
        } else if (a == 2) {
            count = (int) (2 * Math.pow(4, n - 1));
        } else {
            count = (int) (Math.pow(4, n - 1) * (a - 1) + dfs(n - 1, b));
        }

        return count;
    }
}

File: 프로그래머스/2/148652. 유사 칸토어 비트열/유사 칸토어 비트열.py
def dfs(n, k):
    if n == 1:
        return "11011"[:k].count("1")

    a, b = divmod(k, 5 ** (n - 1))

    if a <= 1:
        cnt = 4 ** (n - 1) * a + dfs(n - 1, b)

    elif a == 2:
        cnt = 2 * 4 ** (n - 1)

    elif a > 2:
        cnt = 4 ** (n - 1) * (a - 1) + dfs(n - 1, b)

    return cnt


def solution(n, l, r):
    answer = dfs(n, r) - dfs(n, l - 1)
    return answer